File: tests/fixtures/sample.rb
# Sample Ruby file exercising all extractor features.

module Networking
  # Maximum number of open connections allowed.
  MAX_CONNECTIONS = 100
  # Default timeout in seconds.
  DEFAULT_TIMEOUT = 30

  # Logs a message to stdout.
  def log(message)
    puts message
  end

  # Base class with shared connection functionality.
  class Base
    # Initialize with a host address.
    def initialize(host)
      @host = host
      @connected = false
    end

    # Returns a string representation of the object.
    def to_s
      "#{self.class.name}(#{@host})"
    end

    def self.version
      "1.0"
    end

    private

    # Internal helper to validate state.
    def validate_state
      raise "Invalid state" unless @host
    end
  end

  # Manages a single network connection.
  class Connection < Base
    def initialize(host, port = 8080)
      super(host)
      @port = port
    end

    # Establishes the connection.
    def connect
      log("Connecting to #{@host}:#{@port}")
      @connected = true
    end

    def disconnect
      @connected = false
    end

    def connected?
      @connected
    end

    # Nested configuration class.
    class Config
      def initialize(timeout = DEFAULT_TIMEOUT)
        @timeout = timeout
      end

      def valid?
        @timeout > 0
      end
    end
  end

  # A pool of connections.
  class Pool < Connection
    def initialize(host, size = 10)
      super(host)
      @size = size
      @connections = []
    end

    # Acquires a connection from the pool.
    def acquire
      if @connections.empty?
        conn = Connection.new(@host)
        conn.connect
        conn
      else
        @connections.pop
      end
    end

    def release(conn)
      @connections.push(conn)
    end
  end
end

File: tests/fixtures/sample.f90
! Sample Fortran file exercising extractor features.
module networking
    implicit none

    ! Maximum number of retries.
    integer, parameter :: MAX_RETRIES = 3
    ! Default port for connections.
    integer, parameter :: DEFAULT_PORT = 8080

    ! Represents a network endpoint.
    type :: Endpoint
        character(len=256) :: host
        integer :: port
        logical :: connected
    end type Endpoint

    ! Extends Endpoint with pool functionality.
    type, extends(Endpoint) :: PooledEndpoint
        integer :: pool_size
    end type PooledEndpoint

    ! Interface for connectable types.
    interface Connectable
        module procedure connect_endpoint
    end interface Connectable

contains

    ! Logs a message with the given level.
    subroutine log_message(level, message)
        character(len=*), intent(in) :: level
        character(len=*), intent(in) :: message
        print *, '[', trim(level), '] ', trim(message)
    end subroutine log_message

    ! Creates a new endpoint.
    function create_endpoint(host, port) result(ep)
        character(len=*), intent(in) :: host
        integer, intent(in), optional :: port
        type(Endpoint) :: ep

        ep%host = host
        if (present(port)) then
            ep%port = port
        else
            ep%port = DEFAULT_PORT
        end if
        ep%connected = .false.
    end function create_endpoint

    ! Connects an endpoint.
    subroutine connect_endpoint(ep)
        type(Endpoint), intent(inout) :: ep
        call log_message("INFO", "Connecting to " // trim(ep%host))
        ep%connected = .true.
    end subroutine connect_endpoint

    ! Disconnects an endpoint.
    subroutine disconnect_endpoint(ep)
        type(Endpoint), intent(inout) :: ep
        ep%connected = .false.
    end subroutine disconnect_endpoint

    ! Checks if endpoint is connected.
    logical function is_connected(ep)
        type(Endpoint), intent(in) :: ep
        is_connected = ep%connected
    end function is_connected

end module networking

program main
    use networking
    implicit none

    type(Endpoint) :: conn

    conn = create_endpoint("localhost", 8080)
    call connect_endpoint(conn)
    call disconnect_endpoint(conn)
end program main
